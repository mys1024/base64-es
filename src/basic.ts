const textEncoder = new TextEncoder()
const textDecoder = new TextDecoder()

const values = new Uint8Array(128)

const b64Chars = new Uint8Array(64)
// 'A' to 'Z'
for (let char = 65; char <= 90; char++) {
  const value = char - 65
  b64Chars[value] = char
  values[char] = value
}
// 'a' to 'z'
for (let char = 97; char <= 122; char++) {
  const value = char - 97 + 26
  b64Chars[value] = char
  values[char] = value
}
// '0' to '9'
for (let char = 48; char <= 57; char++) {
  const value = char - 48 + 52
  b64Chars[value] = char
  values[char] = value
}
// '+' and '/'
b64Chars[62] = 43
b64Chars[63] = 47
values[43] = 62
values[47] = 63

const b64urlChars = new Uint8Array(b64Chars)
// '-' and '_'
b64urlChars[62] = 45
b64urlChars[63] = 95
values[45] = 62
values[95] = 63

function b64Length(byteCount: number, padding: boolean): number {
  let len = Math.ceil(byteCount / 3) * 4
  if (!padding) {
    switch (byteCount % 3) {
      case 1:
        len -= 2
        break
      case 2:
        len -= 1
    }
  }
  return len
}

export function encode(data: string | Uint8Array, url: boolean, padding: boolean): string {
  if (typeof data === 'string')
    data = textEncoder.encode(data)
  const chars = url ? b64urlChars : b64Chars
  const byteCount = data.length
  const b64Buffer = new Uint8Array(b64Length(byteCount, padding))
  let bi = 0
  let tmp = 0
  for (let i = 0; i < byteCount; i++) {
    switch (i % 3) {
      case 0:
        b64Buffer[bi++] = chars[(data[i] & 0b11111100) >> 2]
        tmp = (data[i] & 0b00000011) << 4
        break
      case 1:
        b64Buffer[bi++] = chars[tmp | ((data[i] & 0b11110000) >> 4)]
        tmp = (data[i] & 0b00001111) << 2
        break
      case 2:
        b64Buffer[bi++] = chars[tmp | ((data[i] & 0b11000000) >> 6)]
        b64Buffer[bi++] = chars[(data[i] & 0b00111111)]
    }
  }
  switch (byteCount % 3) {
    case 1:
      b64Buffer[bi++] = chars[tmp]
      if (padding) {
        b64Buffer[bi++] = 61
        b64Buffer[bi++] = 61
      }
      break
    case 2:
      b64Buffer[bi++] = chars[tmp]
      if (padding)
        b64Buffer[bi++] = 61
  }
  return textDecoder.decode(b64Buffer)
}

export function decode(b64: string): Uint8Array {
  const charCount = b64.length - (b64.endsWith('==') ? 2 : b64.endsWith('=') ? 1 : 0)
  const byteCount = Math.ceil(charCount * 6 / 8) - (charCount % 4 === 0 ? 0 : 1)
  const data = new Uint8Array(byteCount)
  const b64Buffer = textEncoder.encode(b64)
  for (let i = 0; i < charCount; i++) {
    const offset = Math.floor(i / 4) * 3
    const val = values[b64Buffer[i]]
    switch (i % 4) {
      case 0:
        data[offset] |= val << 2
        break
      case 1:
        data[offset] |= val >> 4
        data[offset + 1] |= val << 4
        break
      case 2:
        data[offset + 1] |= val >> 2
        data[offset + 2] |= val << 6
        break
      case 3:
        data[offset + 2] |= val
    }
  }
  return data
}

export function decodeToStr(b64: string): string {
  return textDecoder.decode(decode(b64))
}
