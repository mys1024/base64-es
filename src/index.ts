export * from './base64.js'
export * from './base64url.js'
