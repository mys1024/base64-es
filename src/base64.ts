import { encode as _encode } from "./basic"

export function base64Encode(data: string | Uint8Array, padding = true): string {
  return _encode(data, false, padding)  
}

export {
  decode as base64Decode,
  decodeToStr as base64DecodeToStr,
} from "./basic"
