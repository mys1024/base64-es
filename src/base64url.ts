import { textDecoder, textEncoder } from "./basic"

const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'
const values: Record<string, number> = {}
for (let i = 0; i < chars.length; i++)
  values[chars[i]] = i

function encode(data: string | Uint8Array, padding = false): string {
  if (typeof data === 'string')
    data = textEncoder.encode(data)
  let str = ''
  let tmp = 0
  for (let i = 0; i < data.length; i++) {
    switch (i % 3) {
      case 0:
        str += chars[(data[i] & 0b11111100) >> 2]
        tmp = (data[i] & 0b00000011) << 4
        break
      case 1:
        str += chars[tmp | ((data[i] & 0b11110000) >> 4)]
        tmp = (data[i] & 0b00001111) << 2
        break
      case 2:
        str += chars[tmp | ((data[i] & 0b11000000) >> 6)]
        str += chars[(data[i] & 0b00111111)]
        tmp = 0
        break
    }
  }
  switch (data.length % 3) {
    case 1:
      str += chars[tmp]
      if (padding)
        str += '=='
      break
    case 2:
      str += chars[tmp]
      if (padding)
        str += '='
      break
  }
  return str
}

function decode(str: string): Uint8Array {
  const charCount = str.length - (str.endsWith('==') ? 2 : str.endsWith('=') ? 1 : 0)
  const byteCount = Math.ceil(charCount * 6 / 8) - (charCount % 4 === 0 ? 0 : 1)
  const data = new Uint8Array(byteCount)
  for (let i = 0; i < charCount; i++) {
    const offset = Math.floor(i / 4) * 3
    const val = values[str[i]]
    switch (i % 4) {
      case 0:
        data[offset + 0] |= val << 2
        break
      case 1:
        data[offset + 0] |= val >> 4
        data[offset + 1] |= val << 4
        break
      case 2:
        data[offset + 1] |= val >> 2
        data[offset + 2] |= val << 6
        break
      case 3:
        data[offset + 2] |= val
        break
    }
  }
  return data
}

function decodeToString(str: string): string {
  return textDecoder.decode(decode(str))
}

export const Base64Url = {
  encode,
  decode,
  decodeToString,
}
