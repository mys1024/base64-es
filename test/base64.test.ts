/* spell-checker: disable */

import { describe, expect, it } from 'vitest'
import { base64Encode, base64Decode, base64DecodeToStr } from '../src/base64'

describe.concurrent('base64 encoding', () => {
  it('encode with padding', () => {
    expect(
      base64Encode(new Uint8Array([22])),
    ).toBe('Fg==')
    expect(
      base64Encode(new Uint8Array([77, 203])),
    ).toBe('Tcs=')
    expect(
      base64Encode(new Uint8Array([79, 2, 229])),
    ).toBe('TwLl')
    expect(
      base64Encode(new Uint8Array([80, 171, 243, 128])),
    ).toBe('UKvzgA==')
    expect(
      base64Encode('Hello, world!'),
    ).toBe('SGVsbG8sIHdvcmxkIQ==')
    expect(
      base64Encode('你好，🌏！这是一个处理 Base64 的 ESM 库。'),
    ).toBe('5L2g5aW977yM8J+Mj++8gei/meaYr+S4gOS4quWkhOeQhiBCYXNlNjQg55qEIEVTTSDlupPjgII=')
  })

  it('encode without padding', () => {
    expect(
      base64Encode(new Uint8Array([22]), false),
    ).toBe('Fg')
    expect(
      base64Encode(new Uint8Array([80, 171, 243, 128]), false),
    ).toBe('UKvzgA')
    expect(
      base64Encode('Hello, world!', false),
    ).toBe('SGVsbG8sIHdvcmxkIQ')
  })
})

describe.concurrent('base64 decoding', () => {
  it('decode with padding', () => {
    expect(
      base64Decode('UKvzgA=='),
    ).toEqual(new Uint8Array([80, 171, 243, 128]))
    expect(
      base64DecodeToStr('SGVsbG8sIHdvcmxkIQ=='),
    ).toEqual('Hello, world!')
  })

  it('decode without padding', () => {
    expect(
      base64Decode('UKvzgA'),
    ).toEqual(new Uint8Array([80, 171, 243, 128]))
    expect(
      base64DecodeToStr('SGVsbG8sIHdvcmxkIQ'),
    ).toEqual('Hello, world!')
  })
})
