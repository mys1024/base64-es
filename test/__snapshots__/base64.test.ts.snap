// Vitest Snapshot v1

exports[`Base64 > decode with padding 1`] = `
Uint8Array [
  80,
  171,
  243,
  128,
]
`;

exports[`Base64 > decode with padding 2`] = `
Uint8Array [
  72,
  101,
  108,
  108,
  111,
  44,
  32,
  119,
  111,
  114,
  108,
  100,
  33,
]
`;

exports[`Base64 > decode without padding 1`] = `
Uint8Array [
  80,
  171,
  243,
  128,
]
`;

exports[`Base64 > decode without padding 2`] = `
Uint8Array [
  72,
  101,
  108,
  108,
  111,
  44,
  32,
  119,
  111,
  114,
  108,
  100,
  33,
]
`;

exports[`Base64 > encode with padding 1`] = `"Tcs="`;

exports[`Base64 > encode with padding 2`] = `"UKvzgA=="`;

exports[`Base64 > encode with padding 3`] = `"SGVsbG8sIHdvcmxkIQ=="`;

exports[`Base64 > encode with padding 4`] = `"5L2g5aW977yM8J+Mj++8gei/meaYr+S4gOS4quWkhOeQhiBCYXNlNjQg55qE5bqT44CC"`;

exports[`Base64 > encode without padding 1`] = `"Tcs"`;

exports[`Base64 > encode without padding 2`] = `"UKvzgA"`;

exports[`Base64 > encode without padding 3`] = `"SGVsbG8sIHdvcmxkIQ"`;

exports[`Base64 > encode without padding 4`] = `"5L2g5aW977yM8J+Mj++8gei/meaYr+S4gOS4quWkhOeQhiBCYXNlNjQg55qE5bqT44CC"`;
